### PR TITLE
Modify the volume helper logic.

### PR DESCRIPTION
**Pull request by blackpiglet**
Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Review by shubham-pampattiwar — COMMENTED** [review]
Thanks a bunch @blackpiglet Added a few minor comments. In addition to them I think there is one case that we need to cover:
- If a volume has a matching `snapshot` action and at the same time the user also specifies `defaultVolumesToFSBackups: true` then the snapshot would be skipped by CSI plugin, we need to modify the logic here I believe: https://github.com/vmware-tanzu/velero/blob/49eab81807e3d263210ce0210c9c5120d928077c/pkg/backup/actions/csi/pvc_action.go#L136

**Comment by blackpiglet** [comment]
> Thanks a bunch @blackpiglet Added a few minor comments. In addition to them I think there is one case that we need to cover:
> 
> * If a volume has a matching `snapshot` action and at the same time the user also specifies `defaultVolumesToFSBackups: true` then the snapshot would be skipped by CSI plugin, we need to modify the logic here I believe: https://github.com/vmware-tanzu/velero/blob/49eab81807e3d263210ce0210c9c5120d928077c/pkg/backup/actions/csi/pvc_action.go#L136

Thanks. My understanding of those two filters' priority is a bit different. 
The VolumePolicy matching action has a higher priority than the FS backup opt-out fashion way.

File: pkg/backup/item_backupper.go [hunk 1]
```diff
@@ -385,15 +377,26 @@ func (ib *itemBackupper) executeActions(
...
-				return nil, itemFiles, errors.WithStack(err)
+		if groupResource == kuberesource.PersistentVolumeClaims {
+			if actionName == csiBIAPluginName {
+				snapshotVolume, err := ib.volumeHelperImpl.ShouldPerformSnapshot(obj, kuberesource.PersistentVolumeClaims)
+				if err != nil {
+					return nil, itemFiles, errors.WithStack(err)
+				}
+
+				if !snapshotVolume {
+					log.Info(fmt.Sprintf("skipping csi volume snapshot for PVC %s as it does not fit the volume policy criteria specified by the user for snapshot action", namespace+"/"+name))
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/backup/item_backupper.go:388]: Thanks. Removed this line. The `ShouldPerformSnapshot` function already prints the reason of skipping for each case.

File: pkg/backup/item_backupper.go [hunk 2]
```diff
@@ -385,15 +377,26 @@ func (ib *itemBackupper) executeActions(
...
+					ib.trackSkippedPV(obj, kuberesource.PersistentVolumeClaims, volumeSnapshotApproach, "does not satisfy the criteria for volume policy based snapshot action", log)
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/backup/item_backupper.go:389]: Modify the message.

File: pkg/backup/item_backupper.go [hunk 3]
```diff
@@ -538,35 +547,17 @@ func (ib *itemBackupper) takePVSnapshot(obj runtime.Unstructured, log logrus.Fie
...
+	snapshotVolume, err := ib.volumeHelperImpl.ShouldPerformSnapshot(obj, kuberesource.PersistentVolumes)
+	if err != nil {
+		return err
 	}
 
-	if boolptr.IsSetToFalse(ib.backupRequest.Spec.SnapshotVolumes) {
-		log.Info("Backup has volume snapshots disabled; skipping volume snapshot action.")
-		ib.trackSkippedPV(obj, kuberesource.PersistentVolumes, volumeSnapshotApproach, "backup has volume snapshots disabled", log)
+	if !snapshotVolume {
+		log.Info(fmt.Sprintf("skipping volume snapshot for PV %s as it does not fit the volume policy criteria specified by the user for snapshot action", pv.Name))
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/backup/item_backupper.go:556]: Modified.

File: design/Extend-VolumePolicies-to-support-more-actions.md [hunk 4]
```diff
@@ -86,8 +86,11 @@ volumePolicies:
...
 - Another thing to note is the workflow proposed in this design uses the legacy opt-in/opt-out approach as a fallback option. For instance, the user specifies a VolumePolicy but for a particular volume included in the backup there are no actions(fs-backup/snapshot) matching in the volume policy for that volume, in such a scenario the legacy approach will be used for backing up the particular volume.
+- The relation between the `VolumePolicy` and the backup's legacy parameter `SnapshotVolumes`: 
+  - The `VolumePolicy`'s `snapshot` action matching for volume has higher priority. When there is a `snapshot` action matching for the selected volume, it will be backed by the snapshot way, no matter of the `backup.Spec.SnapshotVolumes` setting.
+  - If there is no `snapshot` action matching the selected volume in the `VolumePolicy`, then the volume will be backed up by `snapshot` way, if the `backup.Spec.SnapshotVolumes` is not set to false.
```
Comment (Lyndon-Li) [inline design/Extend-VolumePolicies-to-support-more-actions.md:93]: Here we may also need to mention the factor from fs-backup

File: pkg/backup/actions/csi/pvc_action.go [hunk 5]
```diff
@@ -64,14 +63,6 @@ func (p *pvcBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
 }
 
 func (p *pvcBackupItemAction) validateBackup(backup velerov1api.Backup) (valid bool) {
-	// Do nothing if volume snapshots have not been requested in this backup
```
Comment (Lyndon-Li) [inline pkg/backup/actions/csi/pvc_action.go:67]: This means all the BIA plugins should not check `backup.Spec.SnapshotVolumes`, we may need to add this to release notes so as to advice the 3rd plugins to remove this check

File: pkg/util/podvolume/pod_volume.go [hunk 6]
```diff
@@ -29,6 +29,32 @@ import (
 	"github.com/vmware-tanzu/velero/pkg/util"
 )
 
+func ShouldPerformFSBackupLegacy(
```
Comment (shubham-pampattiwar) [inline pkg/util/podvolume/pod_volume.go:32]: Lets remove this function as it is not being used.

**Review by shubham-pampattiwar — COMMENTED** [review]
@blackpiglet PR looks good, just added one comment on removal of unused function.

File: pkg/util/podvolume/pod_volume.go (hunk 6 above)
Comment (blackpiglet) [reply pkg/util/podvolume/pod_volume.go:32]: Thanks. Removed.

**Review by Lyndon-Li — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]
/lgtm

**Pull request merged**